test/jdk/jdk/nio/zipfs/TestPosix.java

```diff
@@ -724,4 +717,82 @@ public void testJarFile() throws IOException {
         int rc = JAR_TOOL.run(System.out, System.err, "xvf", JAR_FILE.toString());
         assertEquals(rc, 0, "Return code of jar call is " + rc + " but expected 0");
     }
+
+    /**
+     * Verify that Files.setPosixPermissions does not alter the
+     * 7 first non-permission bits of the 'external file attributes' field.
+     * @throws IOException if an unexpected IOException occurs
+     */
+    @Test
+    public void preserveRemainingBits() throws IOException {
+        /*
+         * The ZIP test vector used here is created using:
+         * % touch hello.txt
+         * % chmod u+s hello.txt     # setuid
+         * % chmod g+s hello.txt     # setgid
+         * % chmod +t hello.txt      # sticky
+         * % zip hello.zip hello.txt
+         * % cat hello.zip | xxd -ps
+         */
+        byte[] zip = HexFormat.of().parseHex("""
+                504b03040a0000000000d994945700000000000000000000000009001c00
+                68656c6c6f2e7478745554090003aa268365aa26836575780b000104f501
+                00000414000000504b01021e030a0000000000d994945700000000000000
+                0000000000090018000000000000000000a48f0000000068656c6c6f2e74
+                78745554050003aa26836575780b000104f50100000414000000504b0506
+                00000000010001004f000000430000000000
+                """.replaceAll("\n",""));
+
+        // Expected bit values of the 'external file attributes' CEN field in the ZIP above
+        String expectedBits = "1000111110100100";
+                            // ^^^^             file type: 1000 (regular file)
+                            //     ^            setuid: ON
+                            //      ^           setgid: ON
+                            //       ^          sticky: ON
+                            //        ^^^^^^^^^ rwxr--r--  (9 bits)
+
+        // Sanity check that 'external file attributes' has the expected value
+        verifyExternalFileAttribute(zip, expectedBits);
+
+
+        Path zipFile = Path.of("preserve-symlink.zip");
+        Files.write(zipFile, zip);
+
+        // Verify that a read/synch roundtrip preserves the external file attributes
+        try (FileSystem fs = FileSystems.newFileSystem(zipFile, ENV_POSIX)) {
+            Path source = fs.getPath("hello.txt");
+            Files.setLastModifiedTime(source, FileTime.from(Instant.now()));
+        }
+        // Updating last modified time should not modify external file attributes
+        verifyExternalFileAttribute(Files.readAllBytes(zipFile), expectedBits);
+
+        // Verify calling Files.setPosixFilePermissions with current permission set
+        try (FileSystem fs = FileSystems.newFileSystem(zipFile, ENV_POSIX)) {
```

Comment:
This should be a separate test IMHO

Reply:
I've separated two tests `setPermissionsShouldPreserveRemainingBits` and `setLastModifiedTimeShouldNotChangeExternalFileAttribute`, both using the support method `assertExternalFileAttributeUnchanged`.

I agree this was much cleaner, since these are in fact separate and independent (but perhaps overlapping) issues.